### PR TITLE
Create empty dynamic memory arrays more efficiently.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: docs/miscellaneous.rst [hunk 1]
```diff
@@ -64,12 +64,15 @@ The position of ``data[4][9].b`` is at ``keccak256(uint256(9) . keccak256(uint25
...
--  0 - 64: scratch space for hashing methods
-- 64 - 96: currently allocated memory size (aka. free memory pointer)
+- ``0x00`` - ``0x3f``: scratch space for hashing methods
+- ``0x40`` - ``0x5f``: currently allocated memory size (aka. free memory pointer)
+- ``0x60`` - ``0x7f``: zero slot
 
-Scratch space can be used between statements (ie. within inline assembly).
+Scratch space can be used between statements (ie. within inline assembly). The zero slot
+is used as initial value for dynamic memory arrays and should never be written to
+(the free memor pointer points to ``0x80`` initially).
```
Comment (axic) [inline docs/miscellaneous.rst]: `memor` -> `memory`

File: libsolidity/codegen/CompilerUtils.cpp [hunk 2]
```diff
@@ -36,11 +38,13 @@ namespace solidity
 
 const unsigned CompilerUtils::dataStartOffset = 4;
 const size_t CompilerUtils::freeMemoryPointer = 64;
+const size_t CompilerUtils::zeroPointer = CompilerUtils::freeMemoryPointer + 0x20;
 const unsigned CompilerUtils::identityContractAddress = 4;
 
 void CompilerUtils::initialiseFreeMemoryPointer()
 {
-	m_context << u256(freeMemoryPointer + 32);
+	solAssert(zeroPointer == CompilerUtils::freeMemoryPointer + 0x20, "");
```
Comment (axic) [inline libsolidity/codegen/CompilerUtils.cpp]: Can we just have a `static_assert` next to the const above?

File: libsolidity/codegen/CompilerUtils.cpp [hunk 3]
```diff
@@ -36,11 +38,13 @@ namespace solidity
...
+	m_context << u256(zeroPointer + 0x20);
```
Comment (axic) [inline libsolidity/codegen/CompilerUtils.cpp]: Can we have another constant for `userMemoryPointer` (perhaps better naming needed)?

**Review by axic — CHANGES_REQUESTED** [review]

File: libsolidity/codegen/CompilerUtils.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/CompilerUtils.cpp]: How would that help? I need to check that at this point here, the value of the free memory pointer does not overlap with the zero memory region.

File: libsolidity/codegen/CompilerUtils.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/codegen/CompilerUtils.cpp]: I don't really understand what you mean, perhaps we should have a quick call about this.

**Comment by axic** [comment]
@chriseth pushed my idea. It can be squashed.

Also need to rebase this.

File: libsolidity/codegen/CompilerUtils.cpp [hunk 4]
```diff
@@ -36,11 +38,17 @@ namespace solidity
...
 const unsigned CompilerUtils::identityContractAddress = 4;
 
+static_assert(CompilerUtils::freeMemoryPointer == 64, "free memory pointer must start at 64");
+static_assert(CompilerUtils::zeroPointer == 96, "zero pointer must be at 96");
+static_assert(CompilerUtils::userMemoryPointer == 128, " user memory must start at or after 128");
+
 void CompilerUtils::initialiseFreeMemoryPointer()
 {
-	m_context << u256(freeMemoryPointer + 32);
+	m_context << u256(zeroPointer + 0x20);
```
Comment (axic) [inline libsolidity/codegen/CompilerUtils.cpp]: Should be `userMemoryPointer`

**Review by axic — APPROVED** [review]

**Review by axic — CHANGES_REQUESTED** [review]
Actually I feel we need more tests. Working on one now.

**Review by axic — APPROVED** [review]

**Pull request merged**